### ISSUE TITLE
pp_split: outdated reification code?

### ISSUE THREAD

**Issue by richardleach** [Perl/perl5#18077]
**Description**
_pp_split_ has a block of AV reification code (currently at [pp.c#L6047](https://github.com/Perl/perl5/blob/blead/pp.c#L6047)):
```
	    if (!AvREAL(ary)) {
		I32 i;
		AvREAL_on(ary);
		AvREIFY_off(ary);
		for (i = AvFILLp(ary); i >= 0; i--)
		    AvARRAY(ary)[i] = &PL_sv_undef; /* don't free mere refs */
	    }
```
Non-existent array elements have been represented by **NULL** rather than **&PL_sv_undef** since https://github.com/Perl/perl5/commit/ce0d59fdd1c7d145efdf6bf8da56a259fed483e4, so it seems like the _pp_split_ code is no longer correct. I note though that https://github.com/Perl/perl5/commit/ce0d59fdd1c7d145efdf6bf8da56a259fed483e4 did change _pp_splice_ and _pp_aslice_, so am unsure whether _pp_split_ was an oversight or if the use of **&PL_sv_undef** is actually intentional in this case?

**Proposed Action**
If the current behaviour is incorrect, I'm happy to submit a PR. 

If so, since that code block doesn't seem to be a hot one, would it be okay to just make that block call av_reify? i.e.
```
	    if (!AvREAL(ary)) {
		av_reify(ary);
	    }
```

**Comment by richardleach** [comment]
Looking at it again, av_reify() wouldn't be a like-for-like swap at all, but still, should the elements be set to NULL rather than &PL_sv_undef?

**Comment by iabyn** [comment]
On Thu, Aug 20, 2020 at 01:40:48AM -0700, Richard Leach wrote:
> **Description**
> _pp_split_ has a block of AV reification code (currently at [pp.c#L6047](https://github.com/Perl/perl5/blob/blead/pp.c#L6047)):
> ```
> 	    if (!AvREAL(ary)) {
> 		I32 i;
> 		AvREAL_on(ary);
> 		AvREIFY_off(ary);
> 		for (i = AvFILLp(ary); i >= 0; i--)
> 		    AvARRAY(ary)[i] = &PL_sv_undef; /* don't free mere refs */
> 	    }
> ```
> Non-existent array elements have been represented by **NULL** rather than **&PL_sv_undef** since https://github.com/Perl/perl5/commit/ce0d59fdd1c7d145efdf6bf8da56a259fed483e4, so it seems like the _pp_split_ code is no longer correct. I note though that https://github.com/Perl/perl5/commit/ce0d59fdd1c7d145efdf6bf8da56a259fed483e4 did change _pp_splice_ and _pp_aslice_, so am unsure whether _pp_split_ was an oversight or if the use of **&PL_sv_undef** is actually intentional in this case?

That does indeed look like an oversight.

-- 
Please note that ash-trays are provided for the use of smokers,
whereas the floor is provided for the use of all patrons.
    -- Bill Royston

**Comment by richardleach** [comment]
I'll submit a PR.

Is it suitably portable to use Zero() here rather than an `AvARRAY(ary)[i] = NULL` loop?

**Comment by iabyn** [comment]
On Tue, Aug 25, 2020 at 07:15:52PM -0700, Richard Leach wrote:
> Is it suitably portable to use Zero() here rather than an
> `AvARRAY(ary)[i] = NULL` loop?

Should be

-- 
Overhead, without any fuss, the stars were going out.
    -- Arthur C Clarke

**Comment by richardleach** [comment]
Closed with #18158
